File: src/scripts/protocols/quake.ts
import { InvalidResponseError } from "../lib/response-error"

export const info: ProtocolInfo = {
  id: 'quake',
  name: 'Quake',
  feature: 'query',
  transport: 'udp',
}

export const parseQuakeInfo = (data: string): Record<string, string> => {
  if (!data.startsWith('\\')) {
    throw new InvalidResponseError('no leading backslash')
  }

  const parts = data.substr(1).split('\\')
  const count = parts.length

  const dict: Record<string, string> = {}

  for (let i = 0; i < count; i += 2) {
    const key = parts[i]
    const val = parts[i + 1]
    dict[key] = val
  }

  return dict
}

File: data/protocols/quake.lua
-- protocol = {
--   id        = "quake",
--   name      = "Quake",
--   feature   = "query",
--   transport = "udp",
-- }

-- function query()
-- end

local function parse_info(data)
  local values = {}
  local pos = 1

  while true do
    local s, e, key, val = data:find("^\\([%w_]+)\\([^\\]*)", pos)

    if not key then
        break
    end

    values[key] = val
    pos = e + 1
  end

  return values
end

-- function process(data)
--   error("not implemented")
-- end

return {
  parse_info = parse_info,
}
